onnx/common/visitor.h

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) ONNX Project Contributors
+
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+#include "onnx/common/common.h"
+#include "onnx/onnx_pb.h"
+
+namespace ONNX_NAMESPACE {
+namespace internal {
+
+// Visitor: A readonly visitor class for ONNX Proto objects.
+// This class is restricted to Nodes, Graphs, Attributes, and Functions.
+// The VisitX methods invoke ProcessX, and if that returns true, will
+// continue to visit all children of the X.
+
+struct Visitor {
+  virtual void VisitGraph(const GraphProto& graph) {
+    if (ProcessGraph(graph))
+      for (auto& node : graph.node())
+        VisitNode(node);
+  }
+
+  virtual void VisitFunction(const FunctionProto& function) {
+    if (ProcessFunction(function))
+      for (auto& node : function.node())
+        VisitNode(node);
+  }
+
+  virtual void VisitNode(const NodeProto& node) {
+    if (ProcessNode(node)) {
+      for (auto& attr : node.attribute()) {
+        VisitAttribute(attr);
+      }
+    }
+  }
+
+  virtual void VisitAttribute(const AttributeProto& attr) {
+    if (ProcessAttribute(attr)) {
+      if (attr.has_g()) {
+        VisitGraph(attr.g());
+      }
+      for (auto& graph : attr.graphs())
+        VisitGraph(graph);
+    }
+  }
+
+  virtual bool ProcessGraph(const GraphProto& graph) {
+    ONNX_UNUSED_PARAMETER(graph);
+    return true;
+  }
+
+  virtual bool ProcessFunction(const FunctionProto& function) {
+    ONNX_UNUSED_PARAMETER(function);
+    return true;
+  }
+
+  virtual bool ProcessNode(const NodeProto& node) {
+    ONNX_UNUSED_PARAMETER(node);
+    return true;
+  }
+
+  virtual bool ProcessAttribute(const AttributeProto& attr) {
+    ONNX_UNUSED_PARAMETER(attr);
+    return true;
+  }
+
+  virtual ~Visitor() {}
+};
+
+// MutableVisitor: A version of Visitor that allows mutation of the visited objects.
+struct MutableVisitor {
```

Comment:
The class is very similar to the previous one. Is there a way to factorize the code?

Reply:
I thought of using templates to do it, but it seems a bit complicated and not worth it. First, we need to find a way to parametrize over "const" types (one works with "const Proto" and the other with "non const Proto"). Second, protobuf generated code works differently for accessing "const" and "non-const" fields, so we can't write the same code anyway. The mutable version is more general and can always be used for both purposes, but the readonly version is better documentation for readonly methods.